### ISSUE TITLE
linking problem, undefined symbol 

### ISSUE THREAD

**Issue by laverite** [tuncer/re2#15]
With centos6.5 + devtoolset-2 I am able to compile the re2 nif.  However when I try to run the eunit tests the module fails to load with:

`The on_load function for module re2 returned {error,
                                              {load_failed,
                                               "Failed to load NIF library: '/tmp/re2/_build/test/lib/re2/priv/re2_nif.so: undefined symbol: _ZNKSt8__detail20_Prime_rehash_policy11_M_next_bktEm'"}}`

If I link to stdc++ by changing https://github.com/tuncer/re2/blob/master/rebar.config.script#L178
  
from:

`{"DRV_LDFLAGS", "$DRV_LDFLAGS " ++ Lib} `

  to:

`{"DRV_LDFLAGS", "$DRV_LDFLAGS " ++ Lib ++ " -l" ++ WhichCPPLib()}`

then the module loads successfully and the eunit tests pass.  Am I doing something wrong?

**Comment by tuncer** [comment]
Thanks, I'll look into. I haven't been able to set up 6.5, would 6.8
with devtoolset-3 be okay? It would be great if you can provide steps
to recreate the same environment (including Erlang).

**Comment by tuncer** [comment]
Can you post a `rebar -vv co` build log? `WhichCPPLib()` is in `$LDFLAGS` and also in the template that references `$DRV_LDFLAGS`, so this error is unexpected. I'll try to find a Linux env that fails...

**Comment by laverite** [comment]
I'll have to find out about using 6.8 and devtoolset-3, but here's a quick Docker file setting up an environment in 6.6 that I can replicate the problem in.  Just bash into the container and run:

-  scl enable devtoolset-2 bash
-  git clone https://github.com/tuncer/re2.git
-  rebar3 compile
-  rebar3 eunit  

[re2-issue-15-docker.txt](https://github.com/tuncer/re2/files/433337/re2-issue-15-docker.txt)

Here's the output of rebar(2) -vv co from my original 6.5 env as requested:

[rebar.out.txt](https://github.com/tuncer/re2/files/433350/rebar.out.txt)

**Comment by tuncer** [comment]
Thanks, I'll see if I can reproduce it, though I'm not sure how soon that will be.

Does the following C++ code build and run?
```
#include <vector>
#include <iostream>

int main()
{
    const std::vector<int> numbers = {1, 2, 3, 4};
    for (auto const &num : numbers) {
        std::cout << num << std::endl;
    }
    return 0;
}
```

**Comment by tuncer** [comment]
This svn commit could be the fix: https://gcc.gnu.org/viewcvs/gcc?view=revision&revision=195676. I suppose that's included in devtoolset-2.

**Comment by tuncer** [comment]
Just to be clear, re2 1.2.2 can be used as the last C++03 release (before we had to introduce C++11 to support newer upstream).

**Comment by laverite** [comment]
yeah the test c++ code above compiles and runs on my 6.5 env via `c++ -std=c++11  -o test test.cpp`

**Comment by laverite** [comment]
those changes seem to work as well.

**Comment by tuncer** [comment]
Does it still build, if you keep that modification and remove `-lstdc++` from the LDFLAGS definition?

**Comment by tuncer** [comment]
rebar patch: https://github.com/rebar/rebar/pull/617

**Comment by tuncer** [comment]
Explicit `-lstdc++` should be obsolete in `rebar.config.script` when `c++` is the compiler driver used for linking, and it would probably allow us to remove C++ stdlib detection in `rebar.config.script`. However, we'd introduce a minimum rebar version requirement.

**Comment by tuncer** [comment]
We might actually be able to conditionally rely on the corrected plugin, and only fall back if the port compiler is too old. A simple min version check should suffice. We'd still have to carry compat code for use with older plugin versions, but not breaking the build is more important than a fallback code branch.

**Comment by tuncer** [comment]
rebar 2.6.3 includes the fix. This ticket should probably stay open until the mentioned rebar.config.script changes are done. Maybe even until rebar3 port_compiler has been synced.

**Comment by tuncer** [comment]
Can you verify that the `ldflags-optional` branch works as expected with rebar `2.6.3` for you as well?

**Comment by tuncer** [comment]
Here's the rebar3 port_compiler patch: https://github.com/blt/port_compiler/pull/25.

Please verify that the `ldflags-optional` branch works for you as well when compiled by rebar >= `2.6.3` and rebar3 port_compiler with https://github.com/blt/port_compiler/pull/25.

**Comment by laverite** [comment]
Thanks @tuncer!

rebar 2.6.3 on `ldflags-optional` branch works.

rebar3 on `ldflags-optional` branch using your `sync` branch of pc plugin works as well.

**Comment by tuncer** [comment]
Released 1.5.0. I've also pinged @tsloughter to release rebar3 port_compiler 1.3.0.
